geom/twkb_write.go

```diff
@@ -0,0 +1,535 @@
+	}
+	if w.hasM {
+		ext |= 0x02
+		ext |= byte(EncodeZigZagInt32(int32(w.precM)) << 5)
+	}
+	w.writeByte(ext)
+}
+
+func (w *TWKBWriter) writePoints(numPoints int, coords ...float64) {
+	w.writePointArray(numPoints, coords)
+}
+
+// Convert a given number of points from floating point to integer coordinates.
+// Utilise and update the running memory of the previous reference point.
+// The input coords must contain numPoints * the number of dimensions values.
+func (w *TWKBWriter) writePointArray(numPoints int, coords []float64) {
+	var buf [binary.MaxVarintLen64]byte
+	c := 0
+	for i := 0; i < numPoints; i++ {
+		for d := 0; d < w.dimensions; d++ {
+			fval := coords[c]
+			ival := int64(fval * w.scalings[d])
+			// Compute bounding box.
+			if !w.bboxValid {
+				w.bboxMin[d] = ival
+				w.bboxMax[d] = ival
+			} else if ival < w.bboxMin[d] {
+				w.bboxMin[d] = ival
+			} else if ival > w.bboxMax[d] {
+				w.bboxMax[d] = ival
+			}
+			// Perform coord differencing to find the int value.
+			ival -= w.refpoint[d]
+			n := binary.PutVarint(buf[:], ival)
+
+			w.twkb = append(w.twkb, buf[:n]...)
+			w.refpoint[d] += ival
+			c++
+		}
+		if !w.bboxValid {
+			w.bboxValid = true
+		}
+	}
+}
+
+func (w *TWKBWriter) writeAdditionalHeaders() {
+	// These are written in this order so that the size of the
+	// bbox is included in the size computation.
+	if w.hasBBox {
+		w.writeBBox()
+	}
+	if w.hasSize {
+		w.writeSize()
+	}
+}
+
+func (w *TWKBWriter) writeBBox() {
+	// Store bbox min and delta for each dimension.
+	var buf [twkbMaxDimensions * 2 * binary.MaxVarintLen64]byte
+	n := 0
+	for d := 0; d < w.dimensions; d++ {
+		n += binary.PutVarint(buf[n:], w.bboxMin[d])
+		n += binary.PutVarint(buf[n:], w.bboxMax[d]-w.bboxMin[d])
+	}
+	// Build a new TWKB buffer with the bbox inserted after the header bytes.
+	start := 2
+	if w.hasExt {
+		start++ // Ensure extended precision byte remains before the bbox values.
+	}
+	// Insert the bbox data.
+	var temp []byte
+	temp = append(temp, w.twkb[:start]...)
+	temp = append(temp, buf[:n]...)
+	temp = append(temp, w.twkb[start:]...)
+	w.twkb = temp
+}
+
+func (w *TWKBWriter) writeSize() {
+	// Compute where to store the size data.
+	start := 2
+	if w.hasExt {
+		start++ // Ensure extended precision byte remains before the size value.
+	}
+	numBytes := len(w.twkb) - start
+	if numBytes < 0 {
+		panic("attempt to add size value to buffer lacking TWKB header bytes")
+	}
+
+	var buf [binary.MaxVarintLen64]byte
+	n := binary.PutUvarint(buf[:], uint64(numBytes))
+
+	// Insert the size data.
+	var temp []byte
+	temp = append(temp, w.twkb[:start]...)
+	temp = append(temp, buf[:n]...)
+	temp = append(temp, w.twkb[start:]...)
+	w.twkb = temp
+}
+
+func (w *TWKBWriter) writeIDList(num int) error {
+	if !w.hasIDs {
+		return nil
+	}
+	if num != len(w.idList) {
+		return fmt.Errorf("unexpected ID list length %d, expected %d", len(w.idList), num)
+	}
+	for i := 0; i < num; i++ {
```

Comment:
nit: perhaps the more idiomatic approach is `for _, id := range w.idList { w.writeSignedVarint(id) }`?